extends Node
class_name StateMachine

@export var initial_state: State
var current_state: State
var states: Dictionary = {}

func _ready():
	# RÃ©cupÃ©rer rÃ©fÃ©rence au player (parent de cette StateMachine)
	var player_ref = get_parent()
	
	# RÃ©cupÃ©rer tous les Ã©tats enfants
	for child in get_children():
		if child is State:
			states[child.name.to_lower()] = child
			child.state_machine = self
			child.player = player_ref  # Assigner directement !
			
	# DÃ©marrer avec l'Ã©tat initial
	if initial_state:
		call_deferred("_start_initial_state")
		
	
	
	if initial_state:
		change_state(initial_state.name.to_lower())
	else:
		print("âŒ Aucun Ã©tat initial dÃ©fini!")
		
func _start_initial_state():
	await get_tree().process_frame

	# VÃ©rifier s'il faut dÃ©marrer en mouvement
	if InputHelper.is_moving():
		change_state("walking")
	else:
		change_state(initial_state.name.to_lower())

func _process(delta):
	if current_state:
		current_state.update(delta)

func _physics_process(delta):
	if current_state:
		current_state.physics_update(delta)
		
	if current_state:
		current_state.physics_update(delta)
	else:
		print("Aucun Ã©tat actuel!") 

func change_state(new_state_name: String):
	var new_state = states.get(new_state_name.to_lower())
	
	if not new_state:
		print("âŒ Ã‰tat introuvable: ", new_state_name)
		print("ğŸ“‹ Ã‰tats disponibles: ", states.keys())
		return
	
	if current_state:
		current_state.exit()
	
	current_state = new_state
	current_state.enter()
	
func debug():
	print("ğŸ” Ã‰tats disponibles:", states.keys())
	print("ğŸ” Ã‰tat initial dÃ©fini:", initial_state.name if initial_state else "AUCUN")
